Spigot-Server-Patches/0358-Custom-plugin-advancement-API.patch

```diff
@@ -0,0 +1,199 @@
+         } else {
+             this.postToMainThread(this::reload);
+         }
+diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+index 9fe555986..a1538075a 100644
+--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
++++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+@@ -51,6 +51,7 @@ import org.bukkit.configuration.ConfigurationSection;
+ import org.bukkit.configuration.file.YamlConfiguration;
+ import org.bukkit.configuration.serialization.ConfigurationSerialization;
+ import org.bukkit.conversations.Conversable;
++import org.bukkit.craftbukkit.advancement.CraftAdvancementManager;
+ import org.bukkit.craftbukkit.boss.CraftBossBar;
+ import org.bukkit.craftbukkit.command.VanillaCommandWrapper;
+ import org.bukkit.craftbukkit.entity.CraftPlayer;
+@@ -170,6 +171,7 @@ public final class CraftServer implements Server {
+     private WarningState warningState = WarningState.DEFAULT;
+     private final BooleanWrapper online = new BooleanWrapper();
+     public CraftScoreboardManager scoreboardManager;
++    public CraftAdvancementManager advancementManager = new CraftAdvancementManager(); // Paper - custom advancement API
+     public boolean playerCommandState;
+     private boolean printSaveWarning;
+     private CraftIconCache icon;
+@@ -1738,6 +1740,12 @@ public final class CraftServer implements Server {
+         return scoreboardManager;
+     }
+ 
++    // Paper start - custom advancement API
++    public CraftAdvancementManager getAdvancementManager() {
++        return advancementManager;
++    }
++    // Paper end
++
+     public void checkSaveState() {
+         if (this.playerCommandState || this.printSaveWarning || this.console.autosavePeriod <= 0) {
+             return;
+diff --git a/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementManager.java b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementManager.java
+new file mode 100644
+index 000000000..a2603305e
+--- /dev/null
++++ b/src/main/java/org/bukkit/craftbukkit/advancement/CraftAdvancementManager.java
+@@ -0,0 +1,96 @@
++package org.bukkit.craftbukkit.advancement;
++
++import com.google.common.collect.LinkedListMultimap;
++import com.google.common.collect.ListMultimap;
++import com.google.common.collect.Maps;
++import net.minecraft.server.*;
++import org.bukkit.NamespacedKey;
++import org.bukkit.advancement.AdvancementData;
++import org.bukkit.advancement.AdvancementManager;
++import org.bukkit.craftbukkit.inventory.CraftItemStack;
++import org.bukkit.craftbukkit.util.CraftNamespacedKey;
++import org.bukkit.plugin.Plugin;
++
++import java.util.Map;
++
++public class CraftAdvancementManager implements AdvancementManager {
++    private ListMultimap<Plugin, AdvancementData> pluginAdvancements = LinkedListMultimap.create();
++    
++    public static volatile boolean needsReload = false;
```

Comment:
`volatile`?

Reply:
maybe some plugin will try loading its advancements off the main thread or something... but yeah probably not really needed there